internal/ask/form/aggregation.go

```diff
@@ -0,0 +1,429 @@
+// TextAggregation holds the aggregated text based answers for a single question
+// marked with the Incude in Aggregations tag, orderd by [question_id][answer].
+type TextAggregation map[string]string
+
+// Aggregation holds the various aggregations and stats collected.
+type Aggregation struct {
+	Group Group                    `json:"group" bson:"group"`
+	Count int                      `json:"count" bson:"count"`
+	MC    map[string]MCAggregation `json:"MultipleChoice" bson:"MultipleChoice"` // Capitalization matches widget type MultipleChoice
+}
+
+// Group defines a key for a multiple choice question / answer combo to be used
+// to define slices of submissions to be aggregated.
+type Group struct {
+	ID       string `json:"group_id" bson:"group_id"`
+	Question string `json:"question" bson:"question"`
+	Answer   string `json:"answer" bson:"answer"`
+}
+
+//==============================================================================
+
+// AggregateFormSubmissions retrieves the submissions for a form, groups them then
+// runs aggregations and counts for each one.
+func AggregateFormSubmissions(context interface{}, db *db.DB, id string) (map[string]Aggregation, error) {
+
+	// Group the submissions.
+	groupedSubmissions, err := GroupSubmissions(context, db, id, 0, 0, submission.SearchOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a container for the grouped aggregations.
+	groupAggregations := make(map[string]Aggregation)
+
+	// Loop through the groups of submissions.
+	for group, submissions := range groupedSubmissions {
+
+		// Perform the multiple choice aggregations.
+		mcAggregations, err := MCAggregate(context, db, id, submissions)
+		if err != nil {
+			return nil, err
+		}
+
+		// Including the aggregation of the group that is being aggregated is redundant, remove.
+		for key, agg := range mcAggregations {
+			if agg.Question == group.Question {
+				delete(mcAggregations, key)
+			}
+		}
+
+		// Pack them in an aggregation along with the submission count and group.
+		agg := Aggregation{
+			Group: group,
+			Count: len(submissions),
+			MC:    mcAggregations,
+		}
+
+		// If this is the "all" group, set the key to all as it is not an answer.
+		groupKey := ""
+		if group.Question == "all" {
+			groupKey = "all"
+		}
+
+		// Groups are ultimately based on a chosen answer. We do not have any keys for answers
+		// so hash the answer text for a unique key.
+		if group.Question != "all" {
+			hasher := md5.New()
+			hasher.Write([]byte(group.Answer))
+			groupKey = hex.EncodeToString(hasher.Sum(nil))
+		}
+
+		// Add the aggregation to the map.
+		groupAggregations[groupKey] = agg
+
+	}
+
+	return groupAggregations, nil
+}
+
+//==============================================================================
+
+// SubmissionGroup is a transport that defines the transport structure for a submission group.
+type SubmissionGroup struct {
+	Submissions map[Group][]submission.Submission `json:"submissions" bson:"submissions"`
```

Comment:
I'm a bit confused as to the request. Are you suggesting making another group here Submissions []submission.Submission? 

This object stores a map by group, each of which is a slice of submissions.